concurrency/mutexmap.go

```diff
@@ -58,50 +58,57 @@ func (a *mutexMap[T]) Lock(key T) {
 	a.lock.RLock()
 	mutex, ok := a.items[key]
 	a.lock.RUnlock()
+	if ok {
+		mutex.Lock()
```

Comment:
Don't you need to also move this before the map's unlock?

Reply:
No, we have to release the outer lock before locking the inner lock else we will get a deadlock. If the mutex is already locked, other operations like `DeleteUnlock` would never complete.

Reply:
I'm not so worried about `DeleteUnlock` because it will never be called before the initial Lock has completed and returned. I think the bigger problem is when we have a delete and a concurrent lock again. Then we would actually get two different mutexes controlling the same resource.